core/drivers/firewall/stm32_iac.c

```diff
@@ -0,0 +1,269 @@
+#define _IAC_VERR_MAJREV_MASK	GENMASK_32(7, 4)
+#define _IAC_VERR_MAJREV_SHIFT	4
+
+/* Periph ID per register */
+#define _PERIPH_IDS_PER_REG	32
+
+#define _IAC_FLD_PREP(field, value)	(((uint32_t)(value) << \
+					  (field ## _SHIFT)) & (field ## _MASK))
+#define _IAC_FLD_GET(field, value)	(((uint32_t)(value) & \
+					  (field ## _MASK)) >> \
+					 (field ## _SHIFT))
+
+#define IAC_EXCEPT_MSB_BIT(x)		((x) * _PERIPH_IDS_PER_REG + \
+					 _PERIPH_IDS_PER_REG - 1)
+#define IAC_EXCEPT_LSB_BIT(x)		((x) * _PERIPH_IDS_PER_REG)
+#define IAC_FIRST_ILAC_IN_REG(x)	(__builtin_ffs((x)) - 1)
+#define IAC_ILAC_ID(reg_val, offset)	(IAC_FIRST_ILAC_IN_REG(reg_val) + \
+					 IAC_EXCEPT_LSB_BIT(offset))
+
+/**
+ * struct iac_driver_data - Hardware information on the IAC peripheral
+ *
+ * @version: Peripheral version number
+ * @num_ilac: Number of IAC lines
+ */
+struct iac_driver_data {
+	uint32_t version;
+	uint8_t num_ilac;
+};
+
+/**
+ * struct stm32_iac_platdata - Platform data for the IAC driver
+ *
+ * @irq_chip: Reference to the main IRQ chip of the platform
+ * @base: Virtual base address of the IAC peripheral
+ * @irq: ID of the IAC interrupt
+ */
+struct stm32_iac_platdata {
+	struct itr_chip *irq_chip;
+	vaddr_t base;
+	size_t irq;
+};
+
+struct iac_device {
+	struct stm32_iac_platdata pdata;
+	struct iac_driver_data *ddata;
+	struct itr_handler *itr;
+};
+
+static struct iac_device iac_dev;
+
+static void stm32_iac_get_hwdata(void)
+{
+	struct iac_driver_data *ddata = iac_dev.ddata;
+	vaddr_t base = iac_dev.pdata.base;
+	uint32_t regval = 0;
+
+	regval = io_read32(base + _IAC_HWCFGR1);
+	ddata->num_ilac = _IAC_FLD_GET(_IAC_HWCFGR1_CFG5, regval);
+
+	ddata->version = io_read32(base + _IAC_VERR);
+
+	DMSG("IAC version %"PRIu32".%"PRIu32,
+	     _IAC_FLD_GET(_IAC_VERR_MAJREV, ddata->version),
+	     _IAC_FLD_GET(_IAC_VERR_MINREV, ddata->version));
+
+	DMSG("HW cap: enabled, num ilac:[%"PRIu8"]", ddata->num_ilac);
+}
+
+static TEE_Result stm32_iac_parse_fdt(const void *fdt, int node)
+{
+	struct stm32_iac_platdata *pdata = &iac_dev.pdata;
+	TEE_Result res = TEE_ERROR_GENERIC;
+	struct dt_node_info dt_info = { };
+	struct io_pa_va base = { };
+
+	fdt_fill_device_info(fdt, &dt_info, node);
+	if (dt_info.status == DT_STATUS_DISABLED ||
+	    dt_info.reg == DT_INFO_INVALID_REG)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	pdata->irq_chip = interrupt_get_main_chip();
+	res = interrupt_dt_get(fdt, node, &pdata->irq_chip, &pdata->irq);
+	if (res)
+		return res;
+
+	base.pa = dt_info.reg;
+	pdata->base = io_pa_or_va_secure(&base, 1);
+
+	return TEE_SUCCESS;
+}
+
+static enum itr_return stm32_iac_itr(struct itr_handler *h __unused)
+{
+	struct iac_driver_data *ddata = iac_dev.ddata;
+	vaddr_t base = iac_dev.pdata.base;
+	unsigned int nreg = DIV_ROUND_UP(ddata->num_ilac, _PERIPH_IDS_PER_REG);
+	bool do_panic = false;
+	unsigned int i = 0;
+	uint32_t isr = 0;
+
+	for (i = 0; i < nreg; i++) {
+		uint32_t offset = sizeof(uint32_t) * i;
+		uint8_t tries = 0;
```

Comment:
by the way, label `tries` is confusing. This is a simple bit counter, prefer name it `n` or `j` or like.
I think the loop at line 165 would be simpler to read as:
```c
for (n = 0; n < _PERIPH_IDS_PER_REG; n++) {
   ...
   if (!isr)
       break;
}